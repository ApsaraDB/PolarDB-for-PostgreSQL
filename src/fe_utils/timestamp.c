/*-------------------------------------------------------------------------
 *
 * timestamp.c
 *	  copy from src/backend/utils/adt/timestamp.c but for fronted.
 *
 * Portions Copyright (c) 1996-2018, PostgreSQL Global Development Group
 * Portions Copyright (c) 1994, Regents of the University of California
 * Portions Copyright (c) 2021, Alibaba Group Holding Limited
 *
 *
 * IDENTIFICATION
 *	  src/fe_utils/timestamp.c
 *
 *-------------------------------------------------------------------------
 */
#include <time.h>

#include "postgres.h"

#include "fe_utils/timestamp.h"
#include "utils/datetime.h"

/* copied from timestamp.c */
pg_time_t
timestamptz_to_time_t(TimestampTz t)
{
	pg_time_t	result;

	result = (pg_time_t) (t / USECS_PER_SEC +
						  ((POSTGRES_EPOCH_JDATE - UNIX_EPOCH_JDATE) * SECS_PER_DAY));
	return result;
}

/*
 * Stopgap implementation of timestamptz_to_str that doesn't depend on backend
 * infrastructure.  This will work for timestamps that are within the range
 * of the platform time_t type.  (pg_time_t is compatible except for possibly
 * being wider.)
 *
 * XXX the return value points to a static buffer, so beware of using more
 * than one result value concurrently.
 *
 * XXX: The backend timestamp infrastructure should instead be split out and
 * moved into src/common.  That's a large project though.
 */
const char *
timestamptz_to_str(TimestampTz dt)
{
	static char buf[MAXDATELEN + 1];
	char		ts[MAXDATELEN + 1];
	char		zone[MAXDATELEN + 1];
	time_t		result = (time_t) timestamptz_to_time_t(dt);
	struct tm  *ltime = localtime(&result);

	strftime(ts, sizeof(ts), "%Y-%m-%d %H:%M:%S", ltime);
	strftime(zone, sizeof(zone), "%Z", ltime);

	snprintf(buf, sizeof(buf), "%s.%06d %s",
			 ts, (int) (dt % USECS_PER_SEC), zone);

	return buf;
}
