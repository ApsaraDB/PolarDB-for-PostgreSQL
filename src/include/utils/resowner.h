/*-------------------------------------------------------------------------
 *
 * resowner.h
 *	  POSTGRES resource owner definitions.
 *
 * Query-lifespan resources are tracked by associating them with
 * ResourceOwner objects.  This provides a simple mechanism for ensuring
 * that such resources are freed at the right time.
 * See utils/resowner/README for more info.
 *
 *
 * Portions Copyright (c) 1996-2018, PostgreSQL Global Development Group
 * Portions Copyright (c) 1994, Regents of the University of California
 *
 * src/include/utils/resowner.h
 *
 *-------------------------------------------------------------------------
 */
#ifndef RESOWNER_H
#define RESOWNER_H


/*
 * ResourceOwner objects are an opaque data structure known only within
 * resowner.c.
 */
typedef struct ResourceOwnerData *ResourceOwner;


/*
 * Globally known ResourceOwners
 */
extern PGDLLIMPORT ResourceOwner CurrentResourceOwner;
extern PGDLLIMPORT ResourceOwner CurTransactionResourceOwner;
extern PGDLLIMPORT ResourceOwner TopTransactionResourceOwner;

/*
 * Resource releasing is done in three phases: pre-locks, locks, and
 * post-locks.  The pre-lock phase must release any resources that are
 * visible to other backends (such as pinned buffers); this ensures that
 * when we release a lock that another backend may be waiting on, it will
 * see us as being fully out of our transaction.  The post-lock phase
 * should be used for backend-internal cleanup.
 */
typedef enum
{
	RESOURCE_RELEASE_BEFORE_LOCKS,
	RESOURCE_RELEASE_LOCKS,
	RESOURCE_RELEASE_AFTER_LOCKS
} ResourceReleasePhase;

/*
 *	Dynamically loaded modules can get control during ResourceOwnerRelease
 *	by providing a callback of this form.
 */
typedef void (*ResourceReleaseCallback) (ResourceReleasePhase phase,
										 bool isCommit,
										 bool isTopLevel,
										 void *arg);

/* POLAR px */
typedef void (*ResourceWalkerCallback) (const struct ResourceOwnerData * owner);
/* POLAR end */

/*
 * Functions in resowner.c
 */

/* generic routines */
extern ResourceOwner ResourceOwnerCreate(ResourceOwner parent,
					const char *name);
extern void ResourceOwnerRelease(ResourceOwner owner,
					 ResourceReleasePhase phase,
					 bool isCommit,
					 bool isTopLevel);
extern void ResourceOwnerDelete(ResourceOwner owner);
extern ResourceOwner ResourceOwnerGetParent(ResourceOwner owner);
extern void ResourceOwnerNewParent(ResourceOwner owner,
					   ResourceOwner newparent);
extern void RegisterResourceReleaseCallback(ResourceReleaseCallback callback,
								void *arg);
extern void UnregisterResourceReleaseCallback(ResourceReleaseCallback callback,
								  void *arg);

/* POLAR px */
extern void PxResourceOwnerWalker(ResourceOwner owner,
							ResourceWalkerCallback callback);
extern const char* polar_get_resourceowner_name(ResourceOwner owner);

#endif							/* RESOWNER_H */
