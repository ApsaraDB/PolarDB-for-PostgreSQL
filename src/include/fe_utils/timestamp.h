/*-------------------------------------------------------------------------
 *
 * Simple timestamp functions for frontend code
 *
 * Portions Copyright (c) 1996-2018, PostgreSQL Global Development Group
 * Portions Copyright (c) 1994, Regents of the University of California
 * Portions Copyright (c) 2021, Alibaba Group Holding limited
 *
 * src/include/fe_utils/timestamp.h
 *
 *-------------------------------------------------------------------------
 */
#ifndef TIMESTAMP_FE_H
#define TIMESTAMP_FE_H
#include "datatype/timestamp.h"
#include "pgtime.h"

extern const char *timestamptz_to_str(TimestampTz dt);
extern pg_time_t timestamptz_to_time_t(TimestampTz t);
#endif							/* TIMESTAMP_FE_H */
