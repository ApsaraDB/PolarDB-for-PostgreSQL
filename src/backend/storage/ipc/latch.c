/*-------------------------------------------------------------------------
 *
 * latch.c
 *	  Routines for inter-process latches
 *
 * The latch interface is a reliable replacement for the common pattern of
 * using pg_usleep() or select() to wait until a signal arrives, where the
 * signal handler sets a flag variable.  See latch.h for more information
 * on how to use them.
 *
 * Portions Copyright (c) 1996-2025, PostgreSQL Global Development Group
 * Portions Copyright (c) 1994, Regents of the University of California
 *
 * IDENTIFICATION
 *	  src/backend/storage/ipc/latch.c
 *
 *-------------------------------------------------------------------------
 */
#include "postgres.h"

#include "miscadmin.h"
#include "port/atomics.h"
#include "storage/latch.h"
#include "storage/waiteventset.h"
#include "utils/resowner.h"

/* A common WaitEventSet used to implement WaitLatch() */
static WaitEventSet *LatchWaitSet;

/* The positions of the latch and PM death events in LatchWaitSet */
#define LatchWaitSetLatchPos 0
#define LatchWaitSetPostmasterDeathPos 1

void
InitializeLatchWaitSet(void)
{
	int			latch_pos PG_USED_FOR_ASSERTS_ONLY;

	Assert(LatchWaitSet == NULL);

	/* Set up the WaitEventSet used by WaitLatch(). */
	LatchWaitSet = CreateWaitEventSet(NULL, 2);
	latch_pos = AddWaitEventToSet(LatchWaitSet, WL_LATCH_SET, PGINVALID_SOCKET,
								  MyLatch, NULL);
	Assert(latch_pos == LatchWaitSetLatchPos);

	/*
	 * WaitLatch will modify this to WL_EXIT_ON_PM_DEATH or
	 * WL_POSTMASTER_DEATH on each call.
	 */
	if (IsUnderPostmaster)
	{
		latch_pos = AddWaitEventToSet(LatchWaitSet, WL_EXIT_ON_PM_DEATH,
									  PGINVALID_SOCKET, NULL, NULL);
		Assert(latch_pos == LatchWaitSetPostmasterDeathPos);
	}
}

/*
 * Initialize a process-local latch.
 */
void
InitLatch(Latch *latch)
{
	latch->is_set = false;
	latch->maybe_sleeping = false;
	latch->owner_pid = MyProcPid;
	latch->is_shared = false;

#ifdef WIN32
	latch->event = CreateEvent(NULL, TRUE, FALSE, NULL);
	if (latch->event == NULL)
		elog(ERROR, "CreateEvent failed: error code %lu", GetLastError());
#endif							/* WIN32 */
}

/*
 * Initialize a shared latch that can be set from other processes. The latch
 * is initially owned by no-one; use OwnLatch to associate it with the
 * current process.
 *
 * InitSharedLatch needs to be called in postmaster before forking child
 * processes, usually right after allocating the shared memory block
 * containing the latch with ShmemInitStruct. (The Unix implementation
 * doesn't actually require that, but the Windows one does.) Because of
 * this restriction, we have no concurrency issues to worry about here.
 *
 * Note that other handles created in this module are never marked as
 * inheritable.  Thus we do not need to worry about cleaning up child
 * process references to postmaster-private latches or WaitEventSets.
 */
void
InitSharedLatch(Latch *latch)
{
#ifdef WIN32
	SECURITY_ATTRIBUTES sa;

	/*
	 * Set up security attributes to specify that the events are inherited.
	 */
	ZeroMemory(&sa, sizeof(sa));
	sa.nLength = sizeof(sa);
	sa.bInheritHandle = TRUE;

	latch->event = CreateEvent(&sa, TRUE, FALSE, NULL);
	if (latch->event == NULL)
		elog(ERROR, "CreateEvent failed: error code %lu", GetLastError());
#endif

	latch->is_set = false;
	latch->maybe_sleeping = false;
	latch->owner_pid = 0;
	latch->is_shared = true;
}

/*
 * Associate a shared latch with the current process, allowing it to
 * wait on the latch.
 *
 * Although there is a sanity check for latch-already-owned, we don't do
 * any sort of locking here, meaning that we could fail to detect the error
 * if two processes try to own the same latch at about the same time.  If
 * there is any risk of that, caller must provide an interlock to prevent it.
 */
void
OwnLatch(Latch *latch)
{
	int			owner_pid;

	/* Sanity checks */
	Assert(latch->is_shared);

	owner_pid = latch->owner_pid;
	if (owner_pid != 0)
		elog(PANIC, "latch already owned by PID %d", owner_pid);

	latch->owner_pid = MyProcPid;
}

/*
 * Disown a shared latch currently owned by the current process.
 */
void
DisownLatch(Latch *latch)
{
	Assert(latch->is_shared);
	Assert(latch->owner_pid == MyProcPid);

	latch->owner_pid = 0;
}

/*
 * Wait for a given latch to be set, or for postmaster death, or until timeout
 * is exceeded. 'wakeEvents' is a bitmask that specifies which of those events
 * to wait for. If the latch is already set (and WL_LATCH_SET is given), the
 * function returns immediately.
 *
 * The "timeout" is given in milliseconds. It must be >= 0 if WL_TIMEOUT flag
 * is given.  Although it is declared as "long", we don't actually support
 * timeouts longer than INT_MAX milliseconds.  Note that some extra overhead
 * is incurred when WL_TIMEOUT is given, so avoid using a timeout if possible.
 *
 * The latch must be owned by the current process, ie. it must be a
 * process-local latch initialized with InitLatch, or a shared latch
 * associated with the current process by calling OwnLatch.
 *
 * Returns bit mask indicating which condition(s) caused the wake-up. Note
 * that if multiple wake-up conditions are true, there is no guarantee that
 * we return all of them in one call, but we will return at least one.
 */
int
WaitLatch(Latch *latch, int wakeEvents, long timeout,
		  uint32 wait_event_info)
{
	WaitEvent	event;

	/* Postmaster-managed callers must handle postmaster death somehow. */
	Assert(!IsUnderPostmaster ||
		   (wakeEvents & WL_EXIT_ON_PM_DEATH) ||
		   (wakeEvents & WL_POSTMASTER_DEATH));

	/*
	 * Some callers may have a latch other than MyLatch, or no latch at all,
	 * or want to handle postmaster death differently.  It's cheap to assign
	 * those, so just do it every time.
	 */
	if (!(wakeEvents & WL_LATCH_SET))
		latch = NULL;
	ModifyWaitEvent(LatchWaitSet, LatchWaitSetLatchPos, WL_LATCH_SET, latch);

	if (IsUnderPostmaster)
		ModifyWaitEvent(LatchWaitSet, LatchWaitSetPostmasterDeathPos,
						(wakeEvents & (WL_EXIT_ON_PM_DEATH | WL_POSTMASTER_DEATH)),
						NULL);

	if (WaitEventSetWait(LatchWaitSet,
						 (wakeEvents & WL_TIMEOUT) ? timeout : -1,
						 &event, 1,
						 wait_event_info) == 0)
		return WL_TIMEOUT;
	else
		return event.events;
}

/*
 * Like WaitLatch, but with an extra socket argument for WL_SOCKET_*
 * conditions.
 *
 * When waiting on a socket, EOF and error conditions always cause the socket
 * to be reported as readable/writable/connected, so that the caller can deal
 * with the condition.
 *
 * wakeEvents must include either WL_EXIT_ON_PM_DEATH for automatic exit
 * if the postmaster dies or WL_POSTMASTER_DEATH for a flag set in the
 * return value if the postmaster dies.  The latter is useful for rare cases
 * where some behavior other than immediate exit is needed.
 *
 * NB: These days this is just a wrapper around the WaitEventSet API. When
 * using a latch very frequently, consider creating a longer living
 * WaitEventSet instead; that's more efficient.
 */
int
WaitLatchOrSocket(Latch *latch, int wakeEvents, pgsocket sock,
				  long timeout, uint32 wait_event_info)
{
	int			ret = 0;
	int			rc;
	WaitEvent	event;
	WaitEventSet *set = CreateWaitEventSet(CurrentResourceOwner, 3);

	if (wakeEvents & WL_TIMEOUT)
		Assert(timeout >= 0);
	else
		timeout = -1;

	if (wakeEvents & WL_LATCH_SET)
		AddWaitEventToSet(set, WL_LATCH_SET, PGINVALID_SOCKET,
						  latch, NULL);

	/* Postmaster-managed callers must handle postmaster death somehow. */
	Assert(!IsUnderPostmaster ||
		   (wakeEvents & WL_EXIT_ON_PM_DEATH) ||
		   (wakeEvents & WL_POSTMASTER_DEATH));

	if ((wakeEvents & WL_POSTMASTER_DEATH) && IsUnderPostmaster)
		AddWaitEventToSet(set, WL_POSTMASTER_DEATH, PGINVALID_SOCKET,
						  NULL, NULL);

	if ((wakeEvents & WL_EXIT_ON_PM_DEATH) && IsUnderPostmaster)
		AddWaitEventToSet(set, WL_EXIT_ON_PM_DEATH, PGINVALID_SOCKET,
						  NULL, NULL);

	if (wakeEvents & WL_SOCKET_MASK)
	{
		int			ev;

		ev = wakeEvents & WL_SOCKET_MASK;
		AddWaitEventToSet(set, ev, sock, NULL, NULL);
	}

	rc = WaitEventSetWait(set, timeout, &event, 1, wait_event_info);

	if (rc == 0)
		ret |= WL_TIMEOUT;
	else
	{
		ret |= event.events & (WL_LATCH_SET |
							   WL_POSTMASTER_DEATH |
							   WL_SOCKET_MASK);
	}

	FreeWaitEventSet(set);

	return ret;
}

/*
 * Sets a latch and wakes up anyone waiting on it.
 *
 * This is cheap if the latch is already set, otherwise not so much.
 *
 * NB: when calling this in a signal handler, be sure to save and restore
 * errno around it.  (That's standard practice in most signal handlers, of
 * course, but we used to omit it in handlers that only set a flag.)
 *
 * NB: this function is called from critical sections and signal handlers so
 * throwing an error is not a good idea.
 */
void
SetLatch(Latch *latch)
{
#ifndef WIN32
	pid_t		owner_pid;
#else
	HANDLE		handle;
#endif

	/*
	 * The memory barrier has to be placed here to ensure that any flag
	 * variables possibly changed by this process have been flushed to main
	 * memory, before we check/set is_set.
	 */
	pg_memory_barrier();

	/* Quick exit if already set */
	if (latch->is_set)
		return;

	latch->is_set = true;

	pg_memory_barrier();
	if (!latch->maybe_sleeping)
		return;

#ifndef WIN32

	/*
	 * See if anyone's waiting for the latch. It can be the current process if
	 * we're in a signal handler. We use the self-pipe or SIGURG to ourselves
	 * to wake up WaitEventSetWaitBlock() without races in that case. If it's
	 * another process, send a signal.
	 *
	 * Fetch owner_pid only once, in case the latch is concurrently getting
	 * owned or disowned. XXX: This assumes that pid_t is atomic, which isn't
	 * guaranteed to be true! In practice, the effective range of pid_t fits
	 * in a 32 bit integer, and so should be atomic. In the worst case, we
	 * might end up signaling the wrong process. Even then, you're very
	 * unlucky if a process with that bogus pid exists and belongs to
	 * Postgres; and PG database processes should handle excess SIGUSR1
	 * interrupts without a problem anyhow.
	 *
	 * Another sort of race condition that's possible here is for a new
	 * process to own the latch immediately after we look, so we don't signal
	 * it. This is okay so long as all callers of ResetLatch/WaitLatch follow
	 * the standard coding convention of waiting at the bottom of their loops,
	 * not the top, so that they'll correctly process latch-setting events
	 * that happen before they enter the loop.
	 */
	owner_pid = latch->owner_pid;
	if (owner_pid == 0)
		return;
	else if (owner_pid == MyProcPid)
		WakeupMyProc();
	else
		WakeupOtherProc(owner_pid);

#else

	/*
	 * See if anyone's waiting for the latch. It can be the current process if
	 * we're in a signal handler.
	 *
	 * Use a local variable here just in case somebody changes the event field
	 * concurrently (which really should not happen).
	 */
	handle = latch->event;
	if (handle)
	{
		SetEvent(handle);

		/*
		 * Note that we silently ignore any errors. We might be in a signal
		 * handler or other critical path where it's not safe to call elog().
		 */
	}
#endif
}

/*
 * Clear the latch. Calling WaitLatch after this will sleep, unless
 * the latch is set again before the WaitLatch call.
 */
void
ResetLatch(Latch *latch)
{
	/* Only the owner should reset the latch */
	Assert(latch->owner_pid == MyProcPid);
	Assert(latch->maybe_sleeping == false);

	latch->is_set = false;

	/*
	 * Ensure that the write to is_set gets flushed to main memory before we
	 * examine any flag variables.  Otherwise a concurrent SetLatch might
	 * falsely conclude that it needn't signal us, even though we have missed
	 * seeing some flag updates that SetLatch was supposed to inform us of.
	 */
	pg_memory_barrier();
}
