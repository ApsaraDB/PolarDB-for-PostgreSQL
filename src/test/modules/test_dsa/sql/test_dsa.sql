CREATE EXTENSION test_dsa;

SELECT test_dsa_random(3, 5, 1024, 4096, 'random');
SELECT test_dsa_random(3, 5, 1024, 4096, 'forwards');
SELECT test_dsa_random(3, 5, 1024, 4096, 'backwards');

SELECT count(*) from test_dsa_random_parallel(3, 5, 1024, 8192, 'random', 5);
SELECT count(*) from test_dsa_random_parallel(3, 5, 1024, 8192, 'forwards', 5);
SELECT count(*) from test_dsa_random_parallel(3, 5, 1024, 8192, 'backwards', 5);

SELECT test_dsa_oom();