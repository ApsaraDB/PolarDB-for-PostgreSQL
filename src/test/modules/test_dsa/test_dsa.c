/* -------------------------------------------------------------------------
 *
 * test_dsa.c
 *		Simple exercises for dsa.c.
 *
 * Copyright (C) 2016, PostgreSQL Global Development Group
 *
 * IDENTIFICATION
 *		src/test/modules/test_dsa/test_dsa.c
 *
 * -------------------------------------------------------------------------
 */

#include "postgres.h"

#include "fmgr.h"
#include "funcapi.h"
#include "miscadmin.h"
#include "postmaster/bgworker.h"
#include "storage/latch.h"
#include "storage/proc.h"
#include "utils/builtins.h"
#include "utils/dsa.h"
#include "utils/resowner.h"
#include "utils/timestamp.h"

#include <stdlib.h>
#include <unistd.h>

PG_MODULE_MAGIC;

PG_FUNCTION_INFO_V1(test_dsa_random);
PG_FUNCTION_INFO_V1(test_dsa_random_parallel);
PG_FUNCTION_INFO_V1(test_dsa_oom);

/* Which order to free objects in, within each loop. */
typedef enum
{
	/* Free in random order. */
	MODE_RANDOM,
	/* Free in the same order we allocated (FIFO). */
	MODE_FORWARDS,
	/* Free in reverse order of allocation (LIFO). */
	MODE_BACKWARDS
} test_mode;

/* Per-worker results. */
typedef struct
{
	pid_t pid;
	int64 count;
	int64 elapsed_time_us;
} test_result;

/* Parameters for a test run, passed to workers. */
typedef struct
{
	int loops;
	int num_allocs;
	int min_alloc;
	int max_alloc;
	test_mode mode;
	test_result results[1]; /* indexed by worker number */
} test_parameters;

/* The startup message given to each worker. */
typedef struct
{
	/* How to connect to the shmem area. */
	dsa_handle area_handle;
	/* Where to find the parameters. */
	dsa_pointer parameters;
	/* What index this worker should write results to. */
	Size output_index;
} test_hello;

static test_mode
parse_test_mode(text *mode)
{
	test_mode result = MODE_RANDOM;
	char *cstr = text_to_cstring(mode);

	if (strcmp(cstr, "random") == 0)
		result = MODE_RANDOM;
	else if (strcmp(cstr, "forwards") == 0)
		result = MODE_FORWARDS;
	else if (strcmp(cstr, "backwards") == 0)
		result = MODE_BACKWARDS;
	else
		ereport(ERROR,
				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
				 errmsg("unknown mode")));
	return result;
}

static void
check_parameters(const test_parameters *parameters)
{
	if (parameters->min_alloc < 1 || parameters->min_alloc > parameters->max_alloc)
		ereport(ERROR,
				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
				 errmsg("min_alloc must be >= 1, and min_alloc must be <= max_alloc")));
}

static int
my_tranche_id(void)
{
	static int tranche_id = 0;

	if (tranche_id == 0)
		tranche_id = LWLockNewTrancheId();

	return tranche_id;
}

static int64
timestamp_diff(TimestampTz start, TimestampTz end)
{
	long		secs = 0;
	int			microsecs = 0;

	TimestampDifference(end, start, &secs, &microsecs);
	return secs * 1000 + microsecs;
}

static void
do_random_test(dsa_area *area, Size output_index, test_parameters *parameters)
{
	dsa_pointer *objects;
	int min_alloc;
	int extra_alloc;
	int32 i;
	int32 loop;
	TimestampTz start_time = GetCurrentTimestamp();
	int64 total_allocations = 0;

	/*
	 * Make tests reproducible (on the same computer at least) by using the
	 * same random sequence every time.
	 */
	srand(42);

	min_alloc = parameters->min_alloc;
	extra_alloc = parameters->max_alloc - parameters->min_alloc;

	objects = palloc(sizeof(dsa_pointer) * parameters->num_allocs);
	Assert(objects != NULL);
	for (loop = 0; loop < parameters->loops; ++loop)
	{
		int num_actually_allocated = 0;

		for (i = 0; i < parameters->num_allocs; ++i)
		{
			Size size;
			void *memory;

			/* Adjust size randomly if needed. */
			size = min_alloc;
			if (extra_alloc > 0)
				size += rand() % extra_alloc;

			/* Allocate! */
			objects[i] = dsa_allocate_extended(area, size, DSA_ALLOC_NO_OOM);
			if (!DsaPointerIsValid(objects[i]))
			{
				elog(LOG, "dsa: loop %d: out of memory after allocating %d objects", loop, i + 1);
				break;
			}
			++num_actually_allocated;
			/* Pay the cost of accessing that memory */
			memory = dsa_get_address(area, objects[i]);
			memset(memory, 42, size);
		}
		if (parameters->mode == MODE_RANDOM)
		{
			for (i = 0; i < num_actually_allocated; ++i)
			{
				Size x = rand() % num_actually_allocated;
				Size y = rand() % num_actually_allocated;
				dsa_pointer temp = objects[x];

				objects[x] = objects[y];
				objects[y] = temp;
			}
		}
		if (parameters->mode == MODE_BACKWARDS)
		{
			for (i = num_actually_allocated - 1; i >= 0; --i)
				dsa_free(area, objects[i]);
		}
		else
		{
			for (i = 0; i < num_actually_allocated; ++i)
				dsa_free(area, objects[i]);
		}
		total_allocations += num_actually_allocated;
	}
	pfree(objects);

	parameters->results[output_index].elapsed_time_us =
	 	timestamp_diff(start_time, GetCurrentTimestamp());
	parameters->results[output_index].pid = getpid();
	parameters->results[output_index].count = total_allocations;
}

/* Non-parallel version: just do it. */
Datum
test_dsa_random(PG_FUNCTION_ARGS)
{
	test_parameters parameters;
	dsa_area *area;

	parameters.loops = PG_GETARG_INT32(0);
	parameters.num_allocs = PG_GETARG_INT32(1);
	parameters.min_alloc = PG_GETARG_INT32(2);
	parameters.max_alloc = PG_GETARG_INT32(3);
	parameters.mode = parse_test_mode(PG_GETARG_TEXT_PP(4));
	check_parameters(&parameters);

	area = dsa_create(my_tranche_id());
	do_random_test(area, 0, &parameters);
	dsa_dump(area);
	dsa_detach(area);

	PG_RETURN_NULL();
}

Datum test_dsa_random_worker_main(Datum arg);

Datum
test_dsa_random_worker_main(Datum arg)
{
	test_hello hello;
	dsa_area *area;
	test_parameters *parameters;

	CurrentResourceOwner = ResourceOwnerCreate(NULL, "test_dsa toplevel");

	/* Receive hello message and attach to shmem area. */
	memcpy(&hello, MyBgworkerEntry->bgw_extra, sizeof(hello));
	area = dsa_attach(hello.area_handle);
	Assert(area != NULL);
	parameters = dsa_get_address(area, hello.parameters);
	Assert(parameters != NULL);

	do_random_test(area, hello.output_index, parameters);

	dsa_detach(area);

	return (Datum) 0;
}

/* Parallel version: fork a bunch of background workers to do it. */
Datum
test_dsa_random_parallel(PG_FUNCTION_ARGS)
{
	ReturnSetInfo *rsinfo = (ReturnSetInfo *) fcinfo->resultinfo;
	TupleDesc	tupdesc;
	Tuplestorestate *tupstore;
	MemoryContext per_query_ctx;
	MemoryContext oldcontext;

	test_hello hello;
	test_parameters *parameters;
	dsa_area *area;
	int workers;
	int i;
	BackgroundWorkerHandle **handles;

	/* tuplestore boilerplate stuff... */
	if (rsinfo == NULL || !IsA(rsinfo, ReturnSetInfo))
		ereport(ERROR,
				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
				 errmsg("set-valued function called in context that cannot accept a set")));
	if (!(rsinfo->allowedModes & SFRM_Materialize))
		ereport(ERROR,
				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
				 errmsg("materialize mode required, but it is not " \
						"allowed in this context")));
	if (get_call_result_type(fcinfo, NULL, &tupdesc) != TYPEFUNC_COMPOSITE)
		elog(ERROR, "return type must be a row type");
	per_query_ctx = rsinfo->econtext->ecxt_per_query_memory;
	oldcontext = MemoryContextSwitchTo(per_query_ctx);
	tupstore = tuplestore_begin_heap(true, false, work_mem);
	rsinfo->returnMode = SFRM_Materialize;
	rsinfo->setResult = tupstore;
	rsinfo->setDesc = tupdesc;
	MemoryContextSwitchTo(oldcontext);

	/* Prepare to work! */
	workers = PG_GETARG_INT32(5);
	handles = palloc(sizeof(BackgroundWorkerHandle *) * workers);

	/* Set up the shared memory area. */
	area = dsa_create(my_tranche_id());

	/* The workers then will attach to it. */
	hello.area_handle = dsa_get_handle(area);

	/* Allocate space for the parameters object. */
	hello.parameters = dsa_allocate(area, sizeof(test_parameters) +
									sizeof(test_result) * workers);
	Assert(DsaPointerIsValid(hello.parameters));

	/* Set up the parameters object. */
	parameters = dsa_get_address(area, hello.parameters);
	parameters->loops = PG_GETARG_INT32(0);
	parameters->num_allocs = PG_GETARG_INT32(1);
	parameters->min_alloc = PG_GETARG_INT32(2);
	parameters->max_alloc = PG_GETARG_INT32(3);
	parameters->mode = parse_test_mode(PG_GETARG_TEXT_PP(4));
	check_parameters(parameters);

	/* Start the workers. */
	for (i = 0; i < workers; ++i)
	{
		BackgroundWorker bgw;

		snprintf(bgw.bgw_name, sizeof(bgw.bgw_name), "worker%d", i);
		bgw.bgw_flags = BGWORKER_SHMEM_ACCESS;
		bgw.bgw_start_time = BgWorkerStart_PostmasterStart;
		bgw.bgw_restart_time = BGW_NEVER_RESTART;
		snprintf(bgw.bgw_library_name, sizeof(bgw.bgw_library_name),
				 "test_dsa");
		snprintf(bgw.bgw_function_name, sizeof(bgw.bgw_function_name),
				 "test_dsa_random_worker_main");
		Assert(sizeof(parameters) <= BGW_EXTRALEN);
		/* Each worker will write its output to a different slot. */
		hello.output_index = i;
		memcpy(bgw.bgw_extra, &hello, sizeof(hello));
		bgw.bgw_notify_pid = MyProcPid;

		if (!RegisterDynamicBackgroundWorker(&bgw, &handles[i]))
			elog(ERROR, "Can't start worker");
	}

	/* Wait for the workers to complete. */
	for (i = 0; i < workers; ++i)
		/* erm, should really check for BGWH_STOPPED */
		WaitForBackgroundWorkerShutdown(handles[i]);

	/* Generate result tuples. */
	for (i = 0; i < workers; ++i)
	{
		Datum values[3];
		bool nulls[] = { false, false, false };
		Interval *interval = palloc(sizeof(Interval));

		interval->month = 0;
		interval->day = 0;
		interval->time = parameters->results[i].elapsed_time_us
#ifndef HAVE_INT64_TIMESTAMP
			/ 1000000.0
#endif
			;

		values[0] = Int32GetDatum(parameters->results[i].pid);
		values[1] = Int64GetDatum(parameters->results[i].count);
		values[2] = PointerGetDatum(interval);
		tuplestore_putvalues(tupstore, tupdesc, values, nulls);
	}
	tuplestore_donestoring(tupstore);

	pfree(handles);
	dsa_detach(area);

	return (Datum) 0;
}

/* OOM test. Allocate memory until oom, than free and try allocate again. */
Datum
test_dsa_oom(PG_FUNCTION_ARGS)
{
	test_parameters parameters;
	dsa_area *area;

	parameters.loops = 1;
	parameters.num_allocs = 256;
	parameters.min_alloc = 4096;
	parameters.max_alloc = 4096;
	parameters.mode = MODE_RANDOM;
	check_parameters(&parameters);

	area = dsa_create(my_tranche_id());
	dsa_set_size_limit(area, 1*1024*1024);
	dsa_dump(area);

	do_random_test(area, 0, &parameters);
	dsa_dump(area);
	elog(LOG, "Allocated %ld", parameters.results[0].count);

	/* try again */
	do_random_test(area, 0, &parameters);
	dsa_dump(area);
	elog(LOG, "Allocated %ld", parameters.results[0].count);

	parameters.num_allocs = 4000;
	parameters.min_alloc = 256;
	parameters.max_alloc = 256*1024;
	parameters.mode = MODE_RANDOM;
	check_parameters(&parameters);

	do_random_test(area, 0, &parameters);
	dsa_dump(area);
	elog(LOG, "Allocated %ld", parameters.results[0].count);

	do_random_test(area, 0, &parameters);
	dsa_dump(area);
	elog(LOG, "Allocated %ld", parameters.results[0].count);

	dsa_detach(area);

	PG_RETURN_NULL();
}