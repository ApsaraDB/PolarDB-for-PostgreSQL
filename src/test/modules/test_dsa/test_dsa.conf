polar_enable_shm_aset=on