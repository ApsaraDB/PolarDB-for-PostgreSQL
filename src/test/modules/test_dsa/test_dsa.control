# dsa_test extension
comment = 'Tests for DSA'
default_version = '1.0'
module_pathname = '$libdir/test_dsa'
relocatable = true
