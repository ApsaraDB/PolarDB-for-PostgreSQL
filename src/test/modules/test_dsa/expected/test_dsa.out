CREATE EXTENSION test_dsa;
SELECT test_dsa_random(3, 5, 1024, 4096, 'random');
 test_dsa_random 
-----------------
 
(1 row)

SELECT test_dsa_random(3, 5, 1024, 4096, 'forwards');
 test_dsa_random 
-----------------
 
(1 row)

SELECT test_dsa_random(3, 5, 1024, 4096, 'backwards');
 test_dsa_random 
-----------------
 
(1 row)

SELECT count(*) from test_dsa_random_parallel(3, 5, 1024, 8192, 'random', 5);
 count 
-------
     5
(1 row)

SELECT count(*) from test_dsa_random_parallel(3, 5, 1024, 8192, 'forwards', 5);
 count 
-------
     5
(1 row)

SELECT count(*) from test_dsa_random_parallel(3, 5, 1024, 8192, 'backwards', 5);
 count 
-------
     5
(1 row)

SELECT test_dsa_oom();
 test_dsa_oom 
--------------
 
(1 row)

